Frameworks/UIKit/NSLayoutManager.mm

```diff
@@ -72,105 +85,107 @@ - (void)__layoutAllText {
                 break;
             }
 
-            CGRect remainingRect = {};
-            CGRect rect = [container lineFragmentRectForProposedRect:CGRectMake(origin.x, origin.y, containerSize.width, lineHeight)
-                                                             atIndex:stringIndex
-                                                    writingDirection:NSWritingDirectionLeftToRight
-                                                       remainingRect:&remainingRect];
+            CGRect remainingRect{};
+            proposedRect.origin = origin;
+            const CGRect rect = [container lineFragmentRectForProposedRect:proposedRect
+                                                                   atIndex:stringIndex
+                                                          writingDirection:NSWritingDirectionLeftToRight
+                                                             remainingRect:&remainingRect];
 
             // Approximate how many characters can fit to keep from re-rendering large amounts of text
-            CFIndex lastIndex = std::min(CTLineGetStringIndexForPosition(line, { drawnWidth + rect.size.width, 0.0f }) + 1, lineEnd);
+            const CFIndex lastIndex = std::min(CTLineGetStringIndexForPosition(line, { drawnWidth + rect.size.width, 0.0f }) + 1, lineEnd);
 
             if (lastIndex == stringIndex) {
                 // Unable to fit any text in the current rect, continue to next
                 if (remainingRect.size.width > 0.0f && stringIndex < lineEnd) {
                     origin = { remainingRect.origin.x, origin.y };
                 } else {
                     origin = { 0.0f, origin.y + lineHeight };
-                    _totalSize.height += lineHeight;
                 }
 
                 continue;
             }
 
-            woc::unique_cf<CGPathRef> path{ CGPathCreateWithRect(CGRectMake(0, 0, rect.size.width, lineHeight), nullptr) };
+            frameDrawRect.size.width = rect.size.width;
+            woc::unique_cf<CGPathRef> path{ CGPathCreateWithRect(frameDrawRect, nullptr) };
             woc::unique_cf<CTFrameRef> frame{
                 CTFramesetterCreateFrame(framesetter.get(), { stringIndex, lastIndex - stringIndex }, path.get(), nullptr)
             };
 
             // Create line to fit as much text as possible in given rect
             CTLineRef fitLine = static_cast<CTLineRef>(CFArrayGetValueAtIndex(CTFrameGetLines(frame.get()), 0));
 
-            CFIndex fitLength = CTLineGetStringRange(fitLine).length;
+            const CFIndex fitLength = CTLineGetStringRange(fitLine).length;
             if (fitLength == 0L) {
                 // Failed to fit any text in the current rect, continue to next
                 if (remainingRect.size.width > 0.0f && stringIndex < lineEnd) {
                     origin = { remainingRect.origin.x, origin.y };
                 } else {
                     origin = { 0.0f, origin.y + lineHeight };
```

Comment:
>+ lineHeight [](start = 46, length = 12)

note that this is not always correct, but I'm not sure how best to deal with it, and this is a good approximation for the time being.

a single character with a particularly large font+fontsize in the _paragraph_ affects this lineHeight metric, which is computed from the height across the whole paragraph, even if the chars in the current line are a smaller font size. we can't just take the height of the line we actually drew, either, because if it's just a period or an underscore, that'll throw off the height measurement. what we actually need here is the height of an 'H' in the largest font+fontsize in this current line, but this is messy to determine :s